Google Analytics tracking pixel

ElijahCaine opened the pull request:
Refs [this CoreOS docs PR](/coreos/docs/pull/913).

In an effort to learn what docs people are reading on GitHub **and** how many people are reading those docs, we are adding a Google Analytics tracking pixel. There is a PR to add those docs to the coreos/docs repo (mentioned above). This Pull Request is being created at the request of @joshix.

This PR accomplishes the following:
1. Adds a script which ensures that the tracking pixel is in all of the docs. Allows for the update of the tracking pixel.
2. Adds the tracking pixel to all docs in the repo (i.e., all files that end in `.md`) by running the script.

**Notes**: The script is almost stateless. If you update any aspect of the tracking pixel line then it will correctly replace the existing lines with the updated one upon running the script again.

**Update**:
- The script has been updated to take one of two sub-commands: `add` and `remove`.
  - `add`: Adds or updates the footer in all docs.
  - `remove`: scrubs the footer from the docs.
- Also added a list of 'ignore' directories.
- added some internal refactoring so a it's bit more readable and less spaghetti-code.

euank commented:
This picked up a file in the `vendor` dir. The script should special-case `vendor` and skip it entirely.

lucab commented:
Deferring, internal discussion ongoing.

ElijahCaine commented:
@lucab @euank ( @joshix ): Updated the script to ignore directories (vendor, .github, etc). The script doesn't remove the pixel from ignored directories, just never adds them.

ElijahCaine commented:
@lucab Bumb. Internal discussion still ongoing?

ElijahCaine commented:
So we had an internal discussion and there are mixed feelings about this. Here's a rundown of the [mostly philosophical and non-technical] *problems* people have with these changes:
- We don't want to distribute this pixel in our releases.
- We aren't super excited about adding tracking to the docs in the first place.
- We should encourage people to read the docs on coreos.com instead of github if our goal is gathering analytics.

The responses to these (in order) are:
- A 'remove' sub-command was added to the script so the analytics tracking-pixel can be scrubbed during releases builds.
- More tracking is *on the whole* not a great thing; I and others agree with that claim. That said, we know from experience that a **lot** of people read the docs on GitHub and getting that data (who's reading what) will help us improve the docs people care about upstream and downstream.
- We can add a badge on the README that says something like "Read these docs on CoreOS.com" to encoruage people to read the docs on the fancy site we made.

Thoughts? Have I captured this adequately? @joshix @robszumski @lucab @euank @barakmich

euank commented:
@ElijahCaine add to the list "private mirrors / clones of the repo will still track".

For example, if I were to mirror this into my company's internal gitlab at `source-gitlab.stealthstartup.io/project-codename/rkt`, that origin will be picked up by this tracking. That isn't the intent of this change, but it is the reality, and I doubt stealthstartup wants us to know they're working with rkt containers nor should we want to know unless they volunteer that.

It might be harder to munge this information into the docs -> website conversion, but I'd definitely prefer that.

lucab commented:
@ElijahCaine yes, thanks for summing this up.

On a related note, Github itself is alredy injecting tracking scripts in rendered pages (see [results](https://github.com/coreos/rkt/graphs/traffic)). Can we perhaps integrate those as external sources in our analytics, or even ask Github to provide a more detailed view for project owners?

ElijahCaine commented:
@lucab I don't have permission to view that particular URL 😜 but I can see your point.

When I asked GitHub about our analytics problem they told me that it was not possible to track data per-page on repositories, just over-arching data like you linked. This broad data is useful in most cases (answering the question "How popular am I?") but doesn't help us figure out where to focus documentation improvement/maintenance efforts.

@euank That is a fair critique. I'll be sure to document somewhere prominent before merging that forks should consider removing the footer (which is easy with the script) to avoid contributing to our analytics report.

That said **if** `gitlab.stealthstartup.io/project-org-codename/rkt` *did* for our project and accidentally start reporting data to our analytics account it wouldn't reveal to us that there was a fork at `gitlab.stealthstartup.io/project-org-codename/rkt` but would dumbly report that people were visiting the analogous page on GitHub. We would see (or think) that employees of Steal Startup Inc. were visiting those pages, but we don't actually have the ability to view the *origin* of the page view (in this case, gitlab.stealthstartup.io).

This is a function of the limits on how [ga-beacon](https://github.com/igrigorik/ga-beacon) is able to report analytics data (tldr, it's basically a hack).

lucab commented:
@ElijahCaine for reference, traffic panel shows per-page total view, per-page unique visitors and query-stripped URL referrers. However, it doesn't contain any user info to cross-reference (IP / useragent / metadata) and it is presented in 14d aggregate (not drillable, as least from the UI).

ElijahCaine commented:
@lucab Ah, that is more useful information than I initially thought. I checked the same endpoint on a repo I **do** have access to and it is seems like it fulfills some of the requirements for what we want this PR to fulfill. I'll ask around and see if we think that what GitHub provides is enough or if we should continue pursuing the Google Analytics route.

Possible reasons to still use Google Analytics:
- Reports data on all pages, not just the top 10 most popular ones.
- Reports data going back much further than 14 days.
- Reports what people do **not** read in addition to what people **do** read.

In the end I'd be comfortable using what GitHub offers, but I can also see this PR reporting useful data. It depends on what people had in mind when they asked me to start this project.

ElijahCaine commented:
@lucab @euank @joshix I would love some feedback on this PR. If there is anything else stopping us from adding the analytics footer I'd like to resolve that ASAP.

**UPDATE**:
- I just pushed a fix for a bug I totally didn't catch which is that my editor (and I think most others) add a newline to the file, but the pixel is added *without* a newline; tldr I added the newline to the tracking pixel to fix a bug.
- More importantly I added two notes in the README / CONTRIBUTING file that suggest people read the docs on coreos.com / alert people to the fact that we use the tracking pixel and that if individuals don't want that they can run the script to remove the footer.

joshix commented on CONTRIBUTING.md:
```diff
@@ -60,6 +60,12 @@ working on your contributions.
 CoreOS project docs should follow the [Documentation style and formatting
 guide](https://github.com/coreos/docs/tree/master/STYLE.md). Thank you for documenting!
 
+*Note*: rkt embeds footers which report data to Google Analytics. We encourage
```
"Some rkt documentation contains a footer that collects data for Google Analytics. The script at `scripts/docs/analytics-footer.py` accepts a `remove` subcommand to delete these footers. It's a good idea to exclude the `Documentation` directory from your git commits to avoid merge conflicts afterwards."

That said, this sounds painful.

joshix commented:
@lucab I think the biggest remaining hangup is the onus on anyone forking the repo to git exclude or ignore /Documentation if they run the provided script to remove the tracking code. Does this complicate things to much for someone who wants to contribute?

We'd like to wire this all the way up with one repo, so let me know if we can move toward merging or if we need to talk more about a different path.

@ElijahCaine You'll need to rebase before we can get there, either way, and I added a review on the note in the README.

lucab commented:
As I mentioned when this PR first showed up, my initial negative feedback was mostly that
> We aren't super excited about adding tracking to the docs in the first place.

In particular, I disagree with the idea that a tracking-pickel reference should be committed to every single markdown file in the source repository of a FLOSS project, even if k8s is going down that road. I have nothing against tracking visitors on our own company website or in some other repo dedicated to user-docs, but I'm concerned about doing it here.

That's why I suggested a couple of alternatives (splitting user docs, importing GH traffic data, asking for more details from GH API), as I fundamentally don't agree with the approach in this PR even if I understand the reasons for this.

ElijahCaine commented:
@lucab What are the exact problems you have with embedding this in the Pull Request?

I understand that there *are* concerns, and that "Because K8s did it we can too" is not valid reasoning, but I would love to know *why* the proposed alternatives (which have their own glaring problems like practicality or lack of features) are better than this implementation.

jonboulle commented:
To be clear, Kubernetes does _not_ do this in the main repo - i.e. if I `git clone github.com/kubernetes/kubernetes`, there's no tracking pixel in my checkout.

@ElijahCaine the issue IMO is still that it's at best unexpected for the source code repository of a FOSS project to contain this if someone is just trying to obtain/work with a copy of the code.

euank commented:
@jonboulle it does; grep for "GENERATED_ANALYTICS" or "kubernetes-site.appspot.com". Pretty much all `.md` docs have it for Kubernetes.

@ElijahCaine I'm still not happy with it because I don't associate browsing code and markdown on github with providing my current activities to Google or other external entities outside of github. I don't associate browsing rendered markdown on a locally cloned repo with giving tracking information to anyone at all.
This violates those assumptions.
I use no-script to block most tracking and will absolutely add this new analytics url to my ublock rules or /etc/hosts blackhole if this is merged.

ElijahCaine commented:
@euank @jonboulle Thank you for the responses. Those are all valid concerns and points brought up.

My follow-up is that in managing a FLOSS project there are always sacrifices that need to be made for the health and longevity of the project. Using GitHub is a good example of this. GitHub is a closed-source tool which many FLOSS projects use heavily, or sometimes exclusively, for development. This flies in the face of what I consider to be a fundamental Richard-Stallman-esque FLOSS ideology: using an entirely FLOSS stack. That said GitHub is a very powerful, and more importantly *popular* tool. As a result we sacrifice that principle, and a bit of tracking data that GH collects, for a better experience long-term.

I'm not saying this addition is as big as our choice of source-code hosting service, but the pros ought to be weight against the cons fairly. With this PR we (CoreOS) collect some data on rendered copies of our project docs to help focus our documentation efforts. The documentation people read and care about is able to be improved, but another service is tracking your patterns. There is good and there is bad.

As you brought up, anybody that truly cares about their privacy has the tools available to them to remain anonymous, and I fully support them doing so. This can be reading the plaintext of the README on their local copy, stripping the pixel using the script, or blocking traffic to the pixel. Is this PR a perfect solution to our problem? No. Does one exist? I would love to find out myself.

jonboulle commented:
@euank ah, thanks for correcting my naivety 

@ElijahCaine we're going to sync up offline and figure out a path forward early next week

ElijahCaine commented:
@joshix Any thoughts on the state of this particular PR?

lucab commented:
I'm closing this for now, we may re-spin a similar discussion in the context of CNCF later.
